File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/VramEffectsStorage.java
package com.spvessel.spacevil;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import com.spvessel.spacevil.Core.IEffect;

final class VramEffectsStorage<TKey, TEffect extends IEffect, TResource extends IVramResource> {

    Map<TKey, Map<TEffect, TResource>> resourceStorage = new HashMap<>();
    List<TKey> flushList = new LinkedList<>();
    Lock storageLocker = new ReentrantLock();

    boolean addForFlushing(TKey item) {
        storageLocker.lock();
        try {
            if (!flushList.contains(item)) {
                flushList.add(item);
                return true;
            }
            return false;
        } finally {
            storageLocker.unlock();
        }
    }

    void flush() {
        storageLocker.lock();
        try {
            for (TKey item : flushList) {
                if (resourceStorage.containsKey(item)) {
                    Map<TEffect, TResource> effects = resourceStorage.get(item);
                    for (Map.Entry<TEffect, TResource> value : effects.entrySet()) {
                        value.getValue().clear();
                    }
                    effects.clear();
                    resourceStorage.remove(item);
                }
            }
            flushList.clear();
        } finally {
            storageLocker.unlock();
        }
    }

    Map<TEffect, TResource> getResources(TKey item) {
        if (resourceStorage.containsKey(item)) {
            return resourceStorage.get(item);
        }
        return null;
    }

    boolean addResource(TKey item, TEffect effect, TResource resource) {
        if (resourceStorage.containsKey(item)) {
            resourceStorage.get(item).put(effect, resource);
        } else {
            Map<TEffect, TResource> effects = new HashMap<>();
            effects.put(effect, resource);
            resourceStorage.put(item, effects);
        }
        return true;
    }

    boolean deleteResource(TKey item, TEffect effect) {
        if (resourceStorage.containsKey(item)) {
            Map<TEffect, TResource> effects = resourceStorage.get(item);
            if (effects.containsKey(effect)) {
                effects.get(effect).clear();
                effects.remove(effect);
                return true;
            }
        }
        return false;
    }

    void clear() {
        for (Map<TEffect, TResource> resource : resourceStorage.values()) {
            for (Map.Entry<TEffect, TResource> value : resource.entrySet()) {
                value.getValue().clear();
            }
            resource.clear();
        }
        resourceStorage.clear();
    }
}

File: CSharp/SpaceVIL/SpaceVIL/Engine/VramEffectsStorage.cs
using System;
using System.Collections.Generic;
using System.Threading;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal sealed class VramEffectsStorage<TKey, TEffect, TResource> where TEffect : IEffect where TResource : IVramResource
    {
        internal Dictionary<TKey, Dictionary<TEffect, TResource>> ResourceStorage = new Dictionary<TKey, Dictionary<TEffect, TResource>>();
        internal List<TKey> FlushList = new List<TKey>();
        internal Object StorageLocker = new Object();

        internal bool AddForFlushing(TKey item)
        {
            Monitor.Enter(StorageLocker);
            try
            {
                if (!FlushList.Contains(item))
                {
                    FlushList.Add(item);
                    return true;
                }
                return false;
            }
            finally
            {
                Monitor.Exit(StorageLocker);
            }
        }

        internal void Flush()
        {
            Monitor.Enter(StorageLocker);
            try
            {
                foreach (var item in FlushList)
                {
                    if (ResourceStorage.ContainsKey(item))
                    {
                        var effects = ResourceStorage[item];
                        foreach (var value in effects)
                        {
                            value.Value.Clear();
                        }
                        effects.Clear();
                        ResourceStorage.Remove(item);
                    }
                }
                FlushList.Clear();
            }
            finally
            {
                Monitor.Exit(StorageLocker);
            }
        }

        internal Dictionary<TEffect, TResource> GetResources(TKey item)
        {
            if (ResourceStorage.ContainsKey(item))
            {
                return ResourceStorage[item];
            }
            return null;
        }

        internal bool AddResource(TKey item, TEffect effect, TResource resource)
        {
            if (ResourceStorage.ContainsKey(item))
            {
                ResourceStorage[item].Add(effect, resource);
            }
            else
            {
                var effects = new Dictionary<TEffect, TResource>();
                effects.Add(effect, resource);
                ResourceStorage.Add(item, effects);
            }
            return true;
        }

        internal bool DeleteResource(TKey item, TEffect effect)
        {
            if (ResourceStorage.ContainsKey(item))
            {
                var effects = ResourceStorage[item];
                if (effects.ContainsKey(effect))
                {
                    effects[effect].Clear();
                    effects.Remove(effect);
                    return true;
                }
            }
            return false;
        }

        internal void Clear()
        {
            foreach (var resource in ResourceStorage.Values)
            {
                foreach (var value in resource)
                {
                    value.Value.Clear();
                }
                resource.Clear();
            }
            ResourceStorage.Clear();
        }
    }
}